src/fluree/db/json_ld/shacl.cljc

```diff
@@ -48,29 +49,70 @@
 
   )
 
+(defn throw-property-shape-exception!
+  [msg]
+  (throw (ex-info (str "SHACL PropertyShape exception - " msg ".")
+                      {:status 400 :error :db/shacl-validation})))
+
 (defn validate-property
-  "Validates a PropertyShape against a set of flakes"
+  "Validates a PropertyShape for a single predicate
+  against a set of flakes"
   [{:keys [min-count max-count node-kind]} p-flakes]
   (let [n (count p-flakes)]
     (when (and min-count
                (> min-count n))
-      (throw (ex-info (str "SHACL PropertyShape exception - sh:minCount of " min-count
-                           " higher than actual count of " n ".")
-                      {:status 400 :error :db/shacl-validation})))
+      (throw-property-shape-exception! (str "sh:minCount of " min-count " higher than actual count of " n)))
     (when (and max-count
                (> n max-count))
-      (throw (ex-info (str "SHACL PropertyShape exception - sh:maxCount of " max-count
-                           " lower than actual count of " n ".")
-                      {:status 400 :error :db/shacl-validation})))))
+      (throw-property-shape-exception! (str "sh:maxCount of " max-count " lower than actual count of " n)))))
+
+(defn validate-pair-property
+  "Validates a PropertyShape that compares values
+  for a pair of predicates."
+  [{:keys [pair-constraint] :as _p-shape} lhs-flakes rhs-flakes]
+  (let [flake-value (fn [flake] [(flake/o flake) (flake/dt flake)])]
+    (case pair-constraint
+
+      (:equals :disjoint) (let [lhs-values (into #{} (map flake-value) lhs-flakes)
+                                rhs-values (into #{} (map flake-value) rhs-flakes)]
```

Comment:
I am uncertain about the role of datatypes in these comparisons, especially for `sh:equals` and `sh:disjoint`. I went ahead and factored them in to the validation for those two rules, requiring identical datatypes for equality, but could be convinced something different here.